stan/math/rev/core/deep_copy_vars.hpp

```diff
@@ -20,8 +20,8 @@ namespace math {
  *  Otherwise it will be moved.
  */
 template <typename Arith, typename = require_arithmetic_t<scalar_type_t<Arith>>>
-inline decltype(auto) deep_copy_vars(Arith&& arg) {
-  return std::forward<Arith>(arg);
+inline const Arith& deep_copy_vars(const Arith& arg) {
+  return arg;
```

Comment:
What's up with this?

Reply:
Oh good catch. I think type deduction was failing in a `std::return_type_t` somewhere and I just changed this and it went to working. Let me revert it and check.

Reply:
When I revert it I get this error:

```
./stan/math/rev/functor/coupled_ode_system.hpp:94:9: error: no matching constructor for initialization of 'std::tuple<decltype(deep_copy_vars(std::declval<double>()))>'
      (aka 'tuple<double &&>')
        local_args_tuple_(deep_copy_vars(args)...),
        ^                 ~~~~~~~~~~~~~~~~~~~~
./stan/math/prim/functor/coupled_ode_system.hpp:110:9: note: in instantiation of member function 'stan::math::coupled_ode_system_impl<false, CosArg1, stan::math::var,
      double>::coupled_ode_system_impl' requested here
      : coupled_ode_system_impl<
...
```

Any ideas?

Reply:
What test is this? In `coupled_ode_system` why don't you just use

```cpp
std::tuple<Args...> local_args_tuple_;
``

```cpp
std::tuple<decltype(deep_copy_vars(std::declval<Args>()))...>
      local_args_tuple_;
``

We use the first in reduce_sum as well don't we?

Reply:
Had a look at this. So what you're suggesting would work, but it makes a full copy of any arithmetic arguments that come in.

`deep_copy_vars` only makes copies of things that are vars. Things that are arithmetic just get passed through.

I guess in the future when we get expressions `deep_copy_vars` (or whatever holds its place) will probably evaluate expressions and do even a third thing.

Whatcha think? Would we also be taking care of business by just including a specialization for references to arithmetic things (along with one to const references)? Since it's only one argument there's not the combinatorics explosion.

Reply:
Is there a way to make `coupled_ode_system` do perfect forwarding itself that will clear this up? So that then:

```
std::tuple<Args...> local_args_tuple;
```

The const ref-ness of things is buried in Args? I'm trying to set this up but I don't know what types to pass to the `coupled_system_ode` constructors:

```
template <typename F, typename T_y0, typename... Args>
struct coupled_ode_system
    : public coupled_ode_system_impl<
          std::is_arithmetic<return_type_t<T_y0, Args...>>::value, F, T_y0,
          Args...> {
  coupled_ode_system(const F& f,
                     const Eigen::Matrix<T_y0, Eigen::Dynamic, 1>& y0,
                     std::ostream* msgs, Args&&... args)
      : coupled_ode_system_impl<
            std::is_arithmetic<return_type_t<T_y0, Args...>>::value, F, T_y0,
    Args...>(f, y0, msgs, std::forward<Args>(args)...) {}   //// <----- Here
};
```

I'm getting errors like: 
```
test/unit/math/rev/functor/coupled_ode_system_test.cpp:48:7: error: no matching constructor for initialization of 'coupled_ode_system<decltype(harm_osc), double,
      std::vector<stan::math::var>, std::vector<double>, std::vector<int> >' (aka 'coupled_ode_system<harm_osc_ode_fun_eigen, double, vector<stan::math::var>,
      vector<double>, vector<int> >')
      system(harm_osc, y0, &msgs, theta, x, x_int);
      ^      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./stan/math/prim/functor/coupled_ode_system.hpp:109:3: note: candidate constructor not viable: no known conversion from 'std::vector<stan::math::var>' to
      'std::vector<stan::math::var, std::allocator<stan::math::var> > &&' for 4th argument
  coupled_ode_system(const F& f,
```

The tests I'm running here by the way are:
```
./runTests.py test/unit/math/rev/functor/coupled_ode_system_test
```

Reply:
Alright I got this to work by separating the class template arguments from the constructor arguments.

The problem with this is that whether or not the storage will be efficient is a function of what ```Args``` someone passes to the `coupled_ode_system` constructor.

@SteveBronder whatcha think? Want to get on a hangouts and work this out (it's the last thing here)?

```
template <typename F, typename T_y0, typename... Args>
struct coupled_ode_system
    : public coupled_ode_system_impl<
          std::is_arithmetic<return_type_t<T_y0, Args...>>::value, F, T_y0,
          Args...> {
  template <typename... FArgs>
  coupled_ode_system(const F& f,
                     const Eigen::Matrix<T_y0, Eigen::Dynamic, 1>& y0,
                     std::ostream* msgs, FArgs&&... args)
      : coupled_ode_system_impl<
            std::is_arithmetic<return_type_t<T_y0, Args...>>::value, F, T_y0,
    Args...>(f, y0, msgs, std::forward<FArgs>(args)...) {}
};
```